When view params change, get a new copy of the domainObject

deeptailor opened the pull request:
We have a bug where, if a user edits a domainObject with more than one view, and then saves and switches the view, the new view will receive an outdated domainObject.

This fixes it by fetching the latest domainObject before switching views.

davetsay commented:
this fixes #2813

akhenry commented on src/ui/router/Browse.js:
```diff
@@ -19,10 +19,13 @@ define([
-                    .getByProviderKey(changed.view);
-                viewObject(browseObject, provider);
+                //get latest version of the browseObject
+                openmct.objects.get(browseObject.identifier).then((newBrowseObject) => {
```
This results in an additional object fetch per path change. The problem here, as I understand it, is that the two route handlers are both being invoked for path changes, but that the `change:params` handler is being invoked first, with the old object. `navigateToPath` will already fetch the new object, so I think the right approach here is to invoke the `change:params` handler at the end of `navigateToPath` and to add a semaphore that is checked in `change:params` that prevents it being called twice.

deeptailor replied:
agree to the part where when path changes we call object fetch twice. The issue will still remain when we are only changing the view, which only fires the change:params event. We will still need to get the latest object when only change:params is invoked.

I think the only way to avoid the double fetch would be to have a semaphore when the original fetch happens (when the change:path event fires) to avoid another fetch when the change:params event fires. When only the change:params event fires, we still have to fetch the latest object

akhenry replied:
@deeptailor I see, didn't realize that was an issue as well. How has this not come up before? In any case, `Browse.js` [is already listening for changes to the navigated object](https://github.com/nasa/openmct/blob/87d838c69094d818e1807404b72f8238de86c691/src/ui/router/Browse.js#L64), but right now it doesn't update the `browseObject` variable. If we just updated `browseObject` in the mutation listener as well then I think we're good?

deeptailor commented:
@akhenry, I changed the fix to update the browse object when the domainobject changes. This keeps the browseObject current

deeptailor commented:
@akhenry, that’s what I did. Just updated the browseObject variable in the watcher

Deep Tailor

> On Mar 30, 2020, at 17:07, Andrew Henry <notifications@github.com> wrote:
> 
> ﻿
> @akhenry commented on this pull request.
> 
> In src/ui/router/Browse.js:
> 
> > @@ -19,10 +19,13 @@ define([
>  
>          openmct.router.on('change:params', function (newParams, oldParams, changed) {
>              if (changed.view && browseObject) {
> -                let provider = openmct
> -                    .objectViews
> -                    .getByProviderKey(changed.view);
> -                viewObject(browseObject, provider);
> +                //get latest version of the browseObject
> +                openmct.objects.get(browseObject.identifier).then((newBrowseObject) => {
> @deeptailor I see, didn't realize that was an issue as well. How has this not come up before? In any case, Browse.js is already listening for changes to the navigated object, but right now it doesn't update the browseObject variable. If we just updated browseObject in the mutation listener as well then I think we're good?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

akhenry commented:
> @akhenry, that’s what I did. Just updated the browseObject variable in the watcher

My bad, sorry. Was looking at stale PR. Which is really weird, GitHub is usually pretty good at keeping them up to date. After I reloaded they showed up -

<img width="802" alt="Screen Shot 2020-03-30 at 5 16 35 PM" src="https://user-images.githubusercontent.com/4215777/77973864-59833800-72aa-11ea-998f-6cb98c28319e.png">

akhenry commented:
## Reviewer Checklist

* Changes appear to address issue? Y
* Appropriate unit tests included? N/A
* Code style and in-line documentation are appropriate? Y
* Commit messages meet standards? Y
